jina/serve/runtimes/worker/batch_queue.py

```diff
@@ -0,0 +1,172 @@
+
+
+async def sleep_then_set(time_seconds: float, event: Event):
+    """Sleep for time_seconds and then set the event
+
+    :param time_seconds: time to sleep
+    :param event: event to set
+    """
+    await asyncio.sleep(time_seconds)
+    event.set()
+
+
+class BatchQueue:
+    """A batch queue that holds the data request and the executor."""
+
+    def __init__(
+        self,
+        executor: BaseExecutor,
+        exec_endpoint: str,
+        args: Any,
+        params: Optional[Dict] = None,
+        preferred_batch_size: int = 4,
+        timeout: int = 10_000,
+        max_batch_size: Optional[int] = None,
+    ) -> None:
+        if params is None:
+            params = dict()
+        self._executor: BaseExecutor = executor
+        self._exec_endpoint: str = exec_endpoint
+        self.args = args
+        self.params = params
+
+        self._preferred_batch_size: int = preferred_batch_size
+        self._timeout: int = timeout
+        self._max_batch_size: Optional[int] = max_batch_size
+        if self._max_batch_size:
+            if self._max_batch_size < self._preferred_batch_size:
+                raise ValueError(
+                    f'Max batch size must be greater than or equal to preferred batch size.'
+                    f'Got max batch size {self._max_batch_size} and preferred'
+                    f'batch size {self._preferred_batch_size}'
+                )
+
+        self._reset()
+
+    def __repr__(self) -> str:
+        return (
+            f'{self.__class__.__name__}({self._executor}, {self._exec_endpoint},'
+            f'preferred_batch_size={self._preferred_batch_size}, timeout={self._timeout})'
+        )
+
+    def __str__(self) -> str:
+        return self.__repr__()
+
+    def _reset(self) -> None:
+        """Set all events and reset the batch queue."""
+        self._requests: List[DataRequest] = []
+        self._request_lens: List[int] = []
+        self._big_doc: DocumentArray = DocumentArray.empty()
+
+        self._flush_trigger: Event = Event()
+        self._flush_task: Task = asyncio.create_task(
+            self.await_then_flush(self._flush_trigger)
+        )
+        self._timer_started: bool = False
+
+    def _start_timer(self):
+        asyncio.create_task(sleep_then_set(self._timeout / 1000, self._flush_trigger))
+        self._timer_started = True
+
+    async def push(self, request: DataRequest) -> Task:
+        """Append request to the queue. Once the request has been processed, the returned task will complete.
+
+        :param request: The request to append to the queue.
+
+        :return: The task that will be set once the request has been processed.
+        """
+        if self._max_batch_size:
+            if len(self._big_doc) + len(request.docs) > self._max_batch_size:
+                self._flush_trigger.set()
+                await self._flush_task
+
+        if not self._timer_started:
+            self._start_timer()
```

Comment:
according to what was suggested above, we can do:
```suggestion
        if not self. _flush_task:
            self._flush_task = asyncio.create_task(
                self.await_then_flush() # also taking in consideration another suggestion to remove parameter
            )
            asyncio.create_task(self.sleep_then_set()) # also taking in consideration another suggestion to remove parameter
```